images/bootstrap/gcs_restapi.sh

```diff
@@ -0,0 +1,129 @@
+    local jwt_header=$(echo -n '{"alg":"RS256","typ":"JWT"}' | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt_claim=$(echo -n '{"iss":"'$sa_email'","scope":"https://www.googleapis.com/auth/cloud-platform","aud":"https://oauth2.googleapis.com/token","exp":'$(($(date +%s) + 3600))',"iat":'$(date +%s)'}' | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt_signature=$(echo -n "$jwt_header.$jwt_claim" | openssl dgst -binary -sha256 -sign <(echo "$sa_key") | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt="$jwt_header.$jwt_claim.$jwt_signature"
+
+    local response=$(curl -s -X POST https://oauth2.googleapis.com/token \
+         -H "Content-Type: application/x-www-form-urlencoded" \
+         -d "grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&assertion=$jwt")
+
+    access_token=$(echo "$response" | jq -r '.access_token')
+
+    if [ -z "$access_token" ]; then
+        echo "Failed to obtain access token. Check your service account key file."
+        exit 1
+    fi
+}
+
+urlencode_path() {
+    local path="$1"
+    echo "$path" | sed 's/\//%2F/g'
+}
+
+# Function to upload a file to Google Cloud Storage
+upload_to_gcs() {
+    local source_file="$1"
+    local bucket_name="$2"
+    local destination_blob=$(urlencode_path "$3")
+    local content_type="application/octet-stream"
+
+    get_access_token
+
+    upload_response=$(curl -X POST \
+      --data-binary @"$source_file" \
+      -H "Authorization: Bearer $access_token" \
+      -H "Content-Type: $content_type" \
+      "https://storage.googleapis.com/upload/storage/v1/b/$bucket_name/o?uploadType=media&name=$destination_blob")
+
+    if echo "$upload_response" | jq -e '.name' > /dev/null; then
+       echo "File $source_file uploaded successfully as $destination_blob"
+       return 0
+    else
+       echo "Upload failed. Response:"
+       echo "$upload_response" | jq '.'
+       return 1
+    fi
+}
+
+# Function to check if a file exists in GCS
+stat_gcs_file() {
+    local bucket_name="$1"
+    local gcs_file_path=$(urlencode_path "$2")
+
+    get_access_token
+
+    stat_response=$(curl -s -X GET \
+      -H "Authorization: Bearer $access_token" \
+      "https://storage.googleapis.com/storage/v1/b/$bucket_name/o/$gcs_file_path")
+
+    if echo "$stat_response" | jq -e '.error' > /dev/null; then
+        return 1
+    else
+        return 0
+    fi
+}
+
+# Function to read the content of a file from GCS
+cat_gcs_file() {
+    local bucket_name="$1"
+    local gcs_file_path=$(urlencode_path "$2")
+
+    get_access_token
+
+    file_content=$(curl -s -X GET \
+      -H "Authorization: Bearer $access_token" \
+      "https://storage.googleapis.com/storage/v1/b/$bucket_name/o/$gcs_file_path?alt=media")
+
+    if [ -z "$file_content" ]; then
+        echo "Error: No content received"
+        return 1
+    else
+        echo "$file_content"
+        return 0
+    fi
+}
+
+# Function to delete a file from GCS
+rm_gcs_file() {
```

Comment:
@chandramerla where do you see a need for removing files from the gcs bucket? An upload to the same path should just overwrite the file right?

Reply:
Actually I'm creating two files in GCS with SHORT_SHA in its file name - [one](https://github.com/kubevirt/project-infra/blob/02f1a34e5df523efe346e078c890f27e85ae9f0a/github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-periodics.yaml#L133) for saving build tag of x86 centos9 and [another](https://github.com/kubevirt/project-infra/blob/02f1a34e5df523efe346e078c890f27e85ae9f0a/github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml#L52) for x86 k8s providers images.  I'm using SHORT_SHA in filename, so that when two job builds (presubmit or postsubmit) are triggered, to avoid conflicts between them.

These files are used in respective s390x job to find the KUBEVIRTCI_TAG used by x86 image, so as to make sure both arch images and multi-arch manifest all have same tag.

So, I'm using this function at the [end of s390x job](https://github.com/kubevirt/project-infra/blob/02f1a34e5df523efe346e078c890f27e85ae9f0a/github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml#L118) to delete the file. P.S: I missed, calling this function in end of s390x job for presubmit, which I will add